### ISSUE TITLE
Staging Server

### ISSUE THREAD

**Issue by ArnoldSmith86** [ArnoldSmith86/virtualtabletop#22]
Right now the main server (virtualtabletop.io) automatically restarts after each commit to `main` and pulls the latest changes. I think that's fine for now.

But I would like a second server (probably same machine with different port) that does the same for a `staging` branch so we can let people test features before they are final (and can't be changed because people use them).

Do you think that is a good approach? Feature branches/forks get merged into `staging` when they are ready. Then they are tested and afterwards get merged into `main`.

**Comment by rogerl50** [comment]
Would this be done for bug fixes as well?

My understanding is that you guys are moving in the direction of having (more and better) automated testing. Perhaps you should (in the long run) just put all changes (not only feature updates) onto `staging` and after testing release it on a regular basis to `main`.

**Comment by ArnoldSmith86** [comment]
Probably. We'll see how many testers we can get on `staging`. If there aren't many, it's probably quite a hassle to route every little change through there.

**Comment by UltimateGeek** [comment]
Sounds good to me. Eventually I think the deployment on the staging & production servers should happen only after a successful build. But setting up a staging branch initially like main, and getting into the habit of promoting all changes initially to staging is a great idea!

**Comment by rogerl50** [comment]
Agreed. Particularly if there is a unit test framework and the devs get into the habit of providing a unit test with each change, be it a bug or a feature.

**Comment by ArnoldSmith86** [comment]
Oh nice, thanks for adding some labels already.

So, I my plans kinda evolved. I think what would be awesome would be this:

1. Every change to `main` is done using a pull request.
2. Whenever a pull request is created, the server automatically clones it and starts a server for it on a new port.
3. If new commits get added, that version automatically updates that server.
4. The `save` directory would need to be linked or copied between the different pull requests because as a beta-tester you probably don't want to upload your games to every server again.

I think this integrates better into the pull request workflow and a single `staging` branch would probably lead to some problems..

**Comment by ArnoldSmith86** [comment]
You probably noticed that I implemented `2` and `3`. The behaviour can be changed of course.
